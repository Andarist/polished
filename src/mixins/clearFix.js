// @flow
import messageHandlers from '../internalHelpers/_messageHandlers'

/**
 * CSS to contain a float (credit to CSSMojo).
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *    ...clearFix(),
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${clearFix()}
 * `
 *
 * // CSS as JS Output
 *
 * '&::after': {
 *   'clear': 'both',
 *   'content': '""',
 *   'display': 'table'
 * }
 */

function clearFix(parent?: string = '&') {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      messageHandlers('mixins/clearFix.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, max: 1 },
        typeChecks: { param: parent, type: 'string' },
      })
    ) {
      return {}
    }
  }

  const pseudoSelector = `${parent}::after`
  return {
    [pseudoSelector]: {
      clear: 'both',
      content: '""',
      display: 'table',
    },
  }
}

export default clearFix
