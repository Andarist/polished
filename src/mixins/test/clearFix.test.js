// @flow
import clearFix from '../clearFix'

describe('clearFix', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should pass parent to pseudo selector', () => {
    expect(clearFix('div')).toMatchSnapshot()
  })

  it('should default to & when no parent is passed', () => {
    expect(clearFix()).toMatchSnapshot()
  })

  it('should throw an error when called with a non-string value', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    clearFix(1)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw a warning when called with more than 1 parameter', () => {
    clearFix('&', 2)
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })

  it('should throw an error and a warning when called with multiple non-string values', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    clearFix(1, 2)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
