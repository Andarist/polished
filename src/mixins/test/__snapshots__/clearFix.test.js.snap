exports[`clearFix should default to & when no parent is passed 1`] = `
Object {
  "&::after": Object {
    "clear": "both",
    "content": "\"\"",
    "display": "table",
  },
}
`;

exports[`clearFix should pass parent to pseudo selector 1`] = `
Object {
  "div::after": Object {
    "clear": "both",
    "content": "\"\"",
    "display": "table",
  },
}
`;

exports[`clearFix should throw a warning when called with more than 1 parameter 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/clearFix.js --%c

clearFix %cexpects a maximum of 1 parameter. However, you passed 2 parameters. These additional parameters were ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#clearFix%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`clearFix should throw an error and a warning when called with multiple non-string values 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/clearFix.js --%c

clearFix %cexpects a parameter of type string. However, you passed 1(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#clearFix%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`clearFix should throw an error and a warning when called with multiple non-string values 2`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/clearFix.js --%c

clearFix %cexpects a maximum of 1 parameter. However, you passed 2 parameters. These additional parameters were ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#clearFix%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`clearFix should throw an error when called with a non-string value 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/clearFix.js --%c

clearFix %cexpects a parameter of type string. However, you passed 1(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#clearFix%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
